widget_driver/lib/src/dependency_provider.dart

```diff
@@ -0,0 +1,143 @@
+import 'utils/runtime_environment_info.dart';
+
+/// Use this class to create/resolve dependencies in your widgets.
+/// This class adds a safe way to create/resolve dependencies during testing where
+/// parent widgets are not forced to provide a mocked instance of your dependency.
+///
+/// The `DependencyProvider` will create the dependency object
+/// according to the provided builder when you are not running tests.
+///
+/// During testing the `DependencyProvider` will first try to create the dependency object using your builder,
+/// but if this fails and throws (because for example no parent widget registered the dependency),
+/// then the `DependencyProvider` will return the registered test default value.
+///
+/// This is how you use this in your widget where you provide dependencies:
+/// Step one: At the bottom of your widget class, create a subclass of `DependencyProvider` and implement the
+/// `registerTestDefaultFallbackValues` method. In this method you register the test default instance.
+/// This should be an empty implementation instance. Like a mock object.
+///
+/// ```dart
+/// class MyServiceProviderWidget extends StatelessWidget {
+///     ...
+/// }
+///
+/// class _DependencyProvider extends DependencyProvider {
+///     @override
+///     void registerTestDefaultFallbackValues() {
+///         registerTestDefaultBuilder<MyService>(() => TestDefaultMyService());
+///     }
+/// }
+///
+/// class TestDefaultMyService extends EmptyDefault implements TestDefaultMyService {}
+/// ```
+///
+/// And then you use this concrete version of `DependencyProvider` in your widget to create that instance.
+/// Like this:
+///
+/// ```dart
+/// class MyServiceProviderWidget extends StatelessWidget {
+///     ...
+///     @override
+///     Widget build(BuildContext context) {
+///         return Provider(
+///             create: _DependencyProvider().get(() => MyService(someStuffFromContext: context.read<SomeStuff>())),
+///             child: child,
+///         );
+///     }
+/// }
+/// ```
+///
+/// Now when you run this normally as a flutter app, then the real `MyService` will be created.
+/// And it will try to real `SomeStuff` out of the build context.
+/// And when you run this during tests and you registered some mock version of `SomeStuff` in the buildContext
+/// then that will be used.
+///
+/// BUT: if you run this during tests, and you never registered any type for `SomeStuff` in the buildContext then
+/// normally this would break your test build and throw an exception. But in this case it will still run.
+/// It just returns the empty test default instance of your `MyService`.
```

Comment:
How about:

Should you want to provide a mock version of `SomeStuff` during testing, then you can just simply provide that object in the buildContext, like you would previously and we will use that one.

BUT: if you run this during tests, and you never registered any type for `SomeStuff` in the buildContext. Normally, this would break your test build and throw an exception. In this case however, we will default back to your test default instance of your `MyService`, provided in the `registerTestDefaultBuilder()` method. Thereby taking away one more thing to consider/remember when writing tests.

Reply:
Sounds good. I updated to documentation with this.
Done